src/server/utils/file.js

```diff
@@ -0,0 +1,150 @@
+import { promises as fs } from "fs";
+import path from "path";
+import slugify from "slugify";
+import { loadFront } from "yaml-front-matter";
+
+export async function getFiles(dir, recursive = true) {
+  const dirents = await fs.readdir(dir, { withFileTypes: true });
+  const files = await Promise.all(
+    dirents.map((dirent) => {
+      const res = path.resolve(dir, dirent.name);
+
+      if (!recursive && dirent.isDirectory()) return [];
+
+      return dirent.isDirectory() ? getFiles(res) : res;
+    })
+  );
+
+  return files.flat();
+}
+
+export async function processMdFile(dir, filePath) {
+  const pathKey = path
+    .relative(dir, path.dirname(filePath))
+    .split(path.sep)
+    .join("/");
+
+  const rawContent = await fs.readFile(filePath, { encoding: "utf8" });
+  const { __content: content, ...metadata } = loadFront(rawContent);
+  const fileName = path.parse(filePath).name;
+  const route = [pathKey, fileName].filter(Boolean).join("/");
+
+  return {
+    filePath,
+    fileName,
+    pathKey,
+    content,
+    metadata,
+    route,
+  };
+}
+
+export async function getMarkdownPages(dir) {
+  const files = await getFiles(dir);
+  const pages = await Promise.all(files.map((f) => processMdFile(dir, f)));
+
+  // Sort pages by pathKey and metadata order
+  pages.sort((a, b) => {
```

Comment:
I like this as a default, but what about giving users to the option provide their own sorting function.